torch/fx/passes/runtime_assert.py

```diff
@@ -222,89 +306,171 @@ def match_symbol(symint, cb):
             # doesn't really have to be in the loop as it will only run once,
             # it just needs to happen right after the placeholders.
             # insert this after placeholders & added sym nodes, and before non-placeholders.
-            if node not in placeholders:
-                last_sym_node = last_placeholder
-                for _ in range(inserted_sym_nodes):
-                    last_sym_node = last_sym_node.next
-                with graph.inserting_before(last_sym_node.next):
-                    add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+            if node == first_non_placeholder:
+                add_runtime_asserts(ras_by_symbol.pop(None, []))  # type: ignore[call-overload]
+
+            # deduplicate asserts already present in graph
+            if node.target in (
+                torch._check,
+                torch.ops.aten._assert_scalar.default,
+            ):
+                if (
+                    node.args[0] == True  # trivial
+                    or (assert_expr := _get_sym_val(node.args[0])) in added_asserts
+                ):
+                    gm.graph.erase_node(node)
+                else:
+                    added_asserts.add(assert_expr)
+
+            # hash cons on symbolic values
+            if (
+                node.op == "call_function"
+                and (sym_expr := _get_sym_val(node)) is not None
+            ):
+                if (
+                    sym_expr in symbol_to_proxy  # example value is redundant
+                    or (
+                        _is_intermediate_tensor_shape_call(node)
+                        and not (sym_expr.free_symbols - symbol_to_proxy.keys())
+                    )  # shape call on intermediate tensor, turn into computation on input shapes
+                ):
+                    if _is_intermediate_tensor_shape_call(node):  # reify from input shapes
+                        symbol_to_proxy[sym_expr] = _sympy_interp(symbol_to_proxy, sym_expr)
+                        # won't try DCE-ing tensor compute here
+                    hash_node = symbol_to_proxy[sym_expr].node
+                    node.replace_all_uses_with(hash_node)
+                    gm.graph.erase_node(node)
+                    log.debug("CSE node %s -> %s for expr %s", node, hash_node, sym_expr)
+                else:
+                    symbol_to_proxy[sym_expr] = fx.Proxy(node)
+
+            if not any(ras for ras in ras_by_symbol.values()):
+                # if no runtime asserts are present, only do CSE,
+                # don't proceed further and add refinement calls
+                continue
+
+            # If the symbol is used, we'll call sym_constrain_range(_for_size) later when we see it anyways,
+            # so delete calls before that
```

Comment:
Hmm, but aren't we obligated to constrain symbols as size, even if they are otherwise not used anywhere in the FX graph?

Reply:
Hm I wasn't sure, in main if there aren't any torch._check_is_size calls, then no sym_constrain_range calls show up in the graph. For example this test case:
```
class M1(torch.nn.Module):
    def forward(self, x1, x2, x3, y):
        z1 = x1.item()
        z2 = x2.item()
        z3 = x3.item()
        torch._check(z1 == (z2 + z3))
        if z2 + z3 == z1:
            return y * 2
        else:
            return y + 3
```

Reply:
I suppose you are threading the needle here. A typically situation something is marked size-like without a `_check_is_size` call in the graph, is if the size is passed into a function which internally, in its meta, checks that it is a size. The near miss here is that, I suppose you would argue, then the size must have been used by a function, so it cannot be dead. I suppose I could still synthetically cause this situation if later I DCE'd the function in question. Another way something can be marked size-like is if you have something that runs in Dynamo that marks the symbol as size-like, but that marking doesn't make it into the graph.

At the end of the day, this is about invariants. If you assume that all size-like variables must be used in the body of the graph, you have forced the rest of the system to actually uphold this invariant. But there isn't really any good reason to require this, and not much more complicated for the pass to do the right thing in the absence of the invariant. The trade seems clear to me.